File: core/src/nl/t64/game/rpg/sfx/ShakeCamera.java
package nl.t64.game.rpg.sfx;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;


public class ShakeCamera {

    private static final float SHAKE_RADIUS = 30f;
    private static final float DIMINISH_THRESHOLD = 2f;
    private static final float DIMINISH_FACTOR = 0.9f;

    private final Vector2 originalPosition;
    private final Vector2 currentPosition;
    private final Vector2 offset;
    private final float originalShakeRadius;

    private float shakeRadius;
    private float randomAngle;
    private boolean isShaking = false;

    public ShakeCamera() {
        this(0f, 0f, SHAKE_RADIUS);
    }

    public ShakeCamera(float shakeRadius) {
        this(0f, 0f, shakeRadius);
    }

    public ShakeCamera(float x, float y) {
        this(x, y, SHAKE_RADIUS);
    }

    public ShakeCamera(float x, float y, float shakeRadius) {
        this.originalPosition = new Vector2(x, y);
        this.currentPosition = new Vector2();
        this.offset = new Vector2();

        this.originalShakeRadius = shakeRadius;
        this.shakeRadius = shakeRadius;
        this.resetShake();
    }

    public void setOriginalPosition(float x, float y) {
        originalPosition.set(x, y);
    }

    public boolean isShaking() {
        return isShaking;
    }

    public void startShaking() {
        isShaking = true;
    }

    public Vector2 getNewShakePosition() {
        computeCameraOffset();
        computeCurrentPosition();
        diminishShake();
        return currentPosition;
    }

    private void computeCameraOffset() {
        offset.x = MathUtils.cosDeg(randomAngle) * shakeRadius;
        offset.y = MathUtils.sinDeg(randomAngle) * shakeRadius;
    }

    private void computeCurrentPosition() {
        currentPosition.set(offset.add(originalPosition));
    }

    private void diminishShake() {
        if (shakeRadius > DIMINISH_THRESHOLD) {
            continueShake();
        } else {
            resetShake();
        }
    }

    private void continueShake() {
        isShaking = true;
        shakeRadius = shakeRadius * DIMINISH_FACTOR;
        seedRandomAngle();
    }

    private void resetShake() {
        isShaking = false;
        shakeRadius = originalShakeRadius;
        seedRandomAngle();
        currentPosition.set(originalPosition);
    }

    private void seedRandomAngle() {
        randomAngle = MathUtils.random(1, 360);
    }

}

File: core/src/nl/t64/game/rpg/sfx/ShakeCamera.kt
package nl.t64.game.rpg.sfx

import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Vector2


private const val SHAKE_RADIUS = 30f
private const val DIMINISH_THRESHOLD = 2f
private const val DIMINISH_FACTOR = 0.9f

class ShakeCamera(x: Float = 0f, y: Float = 0f, aShakeRadius: Float = SHAKE_RADIUS) {

    constructor(x: Float, y: Float) : this(x, y, SHAKE_RADIUS)
    constructor(shakeRadius: Float) : this(0f, 0f, shakeRadius)

    private val originalPosition: Vector2 = Vector2(x, y)
    private val currentPosition: Vector2 = Vector2()
    private val offset: Vector2 = Vector2()
    private val originalShakeRadius: Float = aShakeRadius

    private var shakeRadius: Float = aShakeRadius
    private var randomAngle = 0f

    var isShaking = false

    init {
        resetShake()
    }

    fun setOriginalPosition(x: Float, y: Float) {
        originalPosition.set(x, y)
    }

    fun startShaking() {
        isShaking = true
    }

    fun getNewShakePosition(): Vector2 {
        computeCameraOffset()
        computeCurrentPosition()
        diminishShake()
        return currentPosition
    }

    private fun computeCameraOffset() {
        offset.x = MathUtils.cosDeg(randomAngle) * shakeRadius
        offset.y = MathUtils.sinDeg(randomAngle) * shakeRadius
    }

    private fun computeCurrentPosition() {
        currentPosition.set(offset.add(originalPosition))
    }

    private fun diminishShake() {
        if (shakeRadius > DIMINISH_THRESHOLD) {
            continueShake()
        } else {
            resetShake()
        }
    }

    private fun continueShake() {
        isShaking = true
        shakeRadius *= DIMINISH_FACTOR
        seedRandomAngle()
    }

    private fun resetShake() {
        isShaking = false
        shakeRadius = originalShakeRadius
        seedRandomAngle()
        currentPosition.set(originalPosition)
    }

    private fun seedRandomAngle() {
        randomAngle = MathUtils.random(1, 360).toFloat()
    }

}
